### PR TITLE
Move to Sqlite

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: migrations/00000000000000_diesel_initial_setup/up.sql [hunk 1]
```diff
@@ -47,10 +47,10 @@ CREATE TABLE IF NOT EXISTS transactions (
     epoch BIGINT NOT NULL,
     block_promised BIGINT NOT NULL,
     block_actual BIGINT,
-    signature bytea NOT NULL,
-    validated bool NOT NULL,
+    signature blob NOT NULL,
+    validated INTEGER NOT NULL,
     bundle_id CHAR(43) REFERENCES bundle(id),
-    sent_to_leader bool NOT NULL,
+    sent_to_leader INTEGER NOT NULL,
```
Comment (mikonieminen) [inline migrations/00000000000000_diesel_initial_setup/up.sql]: Please use `BOOLEAN`, it's standard type. Even when it gets translated into integer, it's good indicator for the schema reader about the actual type of this field.

File: migrations/00000000000000_diesel_initial_setup/up.sql [hunk 2]
```diff
@@ -47,10 +47,10 @@ CREATE TABLE IF NOT EXISTS transactions (
     epoch BIGINT NOT NULL,
     block_promised BIGINT NOT NULL,
     block_actual BIGINT,
-    signature bytea NOT NULL,
-    validated bool NOT NULL,
+    signature blob NOT NULL,
```
Comment (mikonieminen) [inline migrations/00000000000000_diesel_initial_setup/up.sql]: Write the field type in all caps as `BLOB`.

File: migrations/00000000000000_diesel_initial_setup/up.sql [hunk 3]
```diff
@@ -47,10 +47,10 @@ CREATE TABLE IF NOT EXISTS transactions (
...
+    validated INTEGER NOT NULL,
```
Comment (mikonieminen) [inline migrations/00000000000000_diesel_initial_setup/up.sql]: Please use BOOLEAN, it's standard type. Even when it gets translated into integer, it's good indicator for the schema reader about the actual type of this field.

File: src/cron/arweave/arweave.rs [hunk 4]
```diff
@@ -48,8 +48,8 @@ pub struct TransactionData {
 #[derive(Deserialize, Serialize, Default, Clone, Debug)]
 pub struct BlockInfo {
     pub id: String,
-    pub timestamp: i64,
-    pub height: i64,
+    pub timestamp: i32,
+    pub height: i32,
```
Comment (mikonieminen) [inline src/cron/arweave/arweave.rs]: Why are these changed? Especially for the timestamp is too small.

File: src/cron/arweave/arweave.rs [hunk 5]
```diff
@@ -115,7 +115,7 @@ pub struct Arweave {
 #[derive(Deserialize, Serialize, Debug)]
 pub struct GqlVariables {
     pub owners: Vec<String>,
-    pub first: i64,
+    pub first: i32,
```
Comment (mikonieminen) [inline src/cron/arweave/arweave.rs]: What is this field used for and why to change it from `i64` to `i32`?

File: src/cron/bundle.rs [hunk 6]
```diff
@@ -30,7 +30,7 @@ pub struct Bundler {
 
 #[derive(Serialize, Deserialize, Default, Debug)]
 pub struct TxReceipt {
-    block: i64,
+    block: i32,
```
Comment (mikonieminen) [inline src/cron/bundle.rs]: `i32` feels a bit small for storing block number. Why is this changed? On the other hand, I would rather use unsigned integers here.

File: src/cron/bundle.rs [hunk 7]
```diff
@@ -222,14 +222,14 @@ where
...
+                        id: Some(receipt.tx_id),
                         epoch: 0, // TODO: implement epoch correctly
                         block_promised: receipt.block,
                         block_actual: current_block,
                         signature: receipt.signature.as_bytes().to_vec(),
-                        validated: true,
+                        validated: 1,
                         bundle_id: Some(bundle_tx.tx_id.clone()),
-                        sent_to_leader: false,
+                        sent_to_leader: 0,
```
Comment (mikonieminen) [inline src/cron/bundle.rs]: I don't really like using `1` and `0` for boolean values, is there no way to use actual boolean values for these fields with SQLite?

File: src/cron/leader.rs [hunk 8]
```diff
@@ -24,7 +24,7 @@ pub struct ValidatorSignature {
 pub struct ReqBody {
     id: String,
     signature: String,
-    block: i64,
+    block: i32,
```
Comment (mikonieminen) [inline src/cron/leader.rs]: Again, feels a bit small field type for block number.

File: src/database/models.rs [hunk 9]
```diff
@@ -5,40 +5,40 @@ use serde::Serialize;
...
+    pub block_height: i32,
 }
 
 #[derive(Serialize, Queryable)]
 pub struct Transaction {
-    pub id: String,
-    pub epoch: i64,
-    pub block_promised: i64,
-    pub block_actual: Option<i64>,
+    pub id: Option<String>,
```
Comment (mikonieminen) [inline src/database/models.rs]: Is this really optional?

File: src/database/schema.rs [hunk 10]
```diff
@@ -1,34 +1,34 @@
 table! {
     bundle (id) {
-        id -> Bpchar,
-        owner_address -> Nullable<Bpchar>,
-        block_height -> Int8,
+        id -> Nullable<Text>,
```
Comment (mikonieminen) [inline src/database/schema.rs]: Why is this column changed to allow `NULL`?

File: src/database/schema.rs [hunk 11]
```diff
@@ -1,34 +1,34 @@
...
+        owner_address -> Nullable<Text>,
```
Comment (mikonieminen) [inline src/database/schema.rs]: Not exactly related to this PR, but I would expect this column not to allow `NULL`? How can we have bundle without owner?

File: src/database/schema.rs [hunk 12]
```diff
@@ -1,34 +1,34 @@
...
+        block_height -> Integer,
     }
 }
 
 table! {
     leaders (address) {
-        address -> Bpchar,
+        address -> Nullable<Text>,
```
Comment (mikonieminen) [inline src/database/schema.rs]: Why is this changed to allow `NULL`?

File: src/database/schema.rs [hunk 13]
```diff
@@ -1,34 +1,34 @@
...
     transactions (id) {
-        id -> Bpchar,
-        epoch -> Int8,
-        block_promised -> Int8,
-        block_actual -> Nullable<Int8>,
-        signature -> Bytea,
-        validated -> Bool,
-        bundle_id -> Nullable<Bpchar>,
-        sent_to_leader -> Bool,
+        id -> Nullable<Text>,
```
Comment (mikonieminen) [inline src/database/schema.rs]: Why is this changed to allow `NULL`?

File: src/database/schema.rs [hunk 14]
```diff
@@ -1,34 +1,34 @@
...
+        epoch -> Integer,
+        block_promised -> Integer,
+        block_actual -> Nullable<Integer>,
+        signature -> Binary,
+        validated -> Integer,
+        bundle_id -> Nullable<Text>,
```
Comment (mikonieminen) [inline src/database/schema.rs:23]: Not exactly related to this PR, but how can we have transaction without bundle ID?

File: src/database/schema.rs [hunk 15]
```diff
@@ -1,34 +1,34 @@
...
+        sent_to_leader -> Integer,
     }
 }
 
 table! {
     validators (address) {
-        address -> Bpchar,
-        url -> Nullable<Varchar>,
+        address -> Nullable<Text>,
```
Comment (mikonieminen) [inline src/database/schema.rs]: Why is this changed to allow `NULL`?

File: src/server/routes/post_tx.rs [hunk 16]
```diff
@@ -41,7 +41,7 @@ pub struct ValidatorSignature {
 pub struct PostTxBody {
     id: String,
     signature: String,
-    block: i64,
+    block: i32,
```
Comment (mikonieminen) [inline src/server/routes/post_tx.rs]: `i32` feels a bit small for block number.

File: src/server/routes/sign.rs [hunk 17]
```diff
@@ -55,7 +55,7 @@ struct SignedBody {
     validator_signature: String,
 }
 
-pub async fn sign_route<Config>(
+pub async fn sign_route<'a, Config>(
```
Comment (mikonieminen) [inline src/server/routes/sign.rs]: What requires this added lifetime parameter? I cannot see any change requiring it.

**Review by mikonieminen — CHANGES_REQUESTED** [review]
There are quite few topics that made me wonder about the changes made in this PR. One additional issue I have with this one is that it adds `unwrap`s when we should work on removing those.

File: migrations/00000000000000_diesel_initial_setup/up.sql (hunk 1 above)
Comment (bluekirby1111) [reply migrations/00000000000000_diesel_initial_setup/up.sql:53]: Sqlite doesn't have boolean type. Any suggestions on this?

https://stackoverflow.com/questions/843780/store-boolean-value-in-sqlite

File: src/cron/arweave/arweave.rs (hunk 5 above)
Comment (ghost) [reply src/cron/arweave/arweave.rs]: Diesel implementations with Sqlite only supports i32

File: src/cron/arweave/arweave.rs (hunk 4 above)
Comment (ghost) [reply src/cron/arweave/arweave.rs]: Ok, I figured out how to use i64 with sqlite

File: src/cron/arweave/arweave.rs [hunk 18]
```diff
@@ -108,7 +108,7 @@ pub enum ArweaveProtocol {
 #[derive(Clone)]
 pub struct Arweave {
     pub host: String,
-    pub port: i32,
+    pub port: i64,
```
Comment (mikonieminen) [inline src/cron/arweave/arweave.rs]: I suppose changing this was an accident. Actually, I would define the port as `u16`.

**Review by mikonieminen — CHANGES_REQUESTED** [review]
Many of my comments were marked resolved without really changing them. I find especially problematic those around database schemas that allow nullable values where to me it looks like broken data if those fields allow `NULL` values. Here is a good article explaining why: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

File: src/database/schema.rs (hunk 14 above)
Comment (ghost) [reply src/database/schema.rs:23]: Yes

It is bundled in intervals, so it may not have a bundle for sometime

Comment (mikonieminen) [reply src/database/schema.rs:23]: OK, thanks for explaining.

File: src/database/models.rs [hunk 19]
```diff
@@ -13,7 +13,7 @@ pub struct Bundle {
 #[derive(Insertable, Clone)]
 #[table_name = "bundle"]
 pub struct NewBundle {
-    pub id: String,
+    pub id: Option<String>,
```
Comment (mikonieminen) [inline src/database/models.rs]: This should not be optional after changing tables.

File: src/database/schema.rs (hunk 11 above)
Comment (mikonieminen) [reply src/database/schema.rs]: Should not be nullable anymore. Weren't these regenerated?

File: src/database/schema.rs (hunk 12 above)
Comment (mikonieminen) [reply src/database/schema.rs]: This field is marked as `PRIMARY KEY`, how can this be nullable?

File: src/database/schema.rs (hunk 15 above)
Comment (mikonieminen) [reply src/database/schema.rs]: This field is marked as `PRIMARY KEY`, how can this be nullable?

**Review by mikonieminen — CHANGES_REQUESTED** [review]

**Comment by mikonieminen** [comment]
There's still something wrong around those nullable fields.

**Comment by ghost** [comment]
It looks like the schema needs to have the fields defined explicitly as NOT NULL (even when it is a primary key) for the Schema.rs file to be generated correctly.

**Comment by ghost** [comment]
Sounds good to me

**Review by mikonieminen — APPROVED** [review]
